ostia-operator/pkg/apicast/apicast.go

```diff
@@ -238,39 +226,145 @@ func newProbe(path string, port int32, initDelay int32, timeout int32, period in
 	}
 }
 
-//TODO - This should returna slice of generic "RateLimit" objects of multiple, mixed types
-func buildRateLimitPolicies(validSvcs []string, limits []v1alpha1.RateLimit) []FixedWindowRateLimiter {
-	var fixedLimitRules []FixedWindowRateLimiter
-	for _, v := range limits {
-		if contains(validSvcs, v.Service) {
-			switch v.Type {
-			case "fixed":
-				if v.TimeUnit == "minutes" {
-					v.TimeInterval = v.TimeInterval * 60
-				}
-				rule := FixedWindowRateLimiter{
-					Count:  v.MaxRequest,
-					Window: v.TimeInterval,
-					Key: LimiterKey{
-						Name:  v.Service,
-						Scope: v.Service,
-					},
-				}
-				fixedLimitRules = append(fixedLimitRules, rule)
-			default:
-				log.Errorf("unknown rate limit type %s for service. ignoring", v.Type, v.Service)
-			}
+func processRateLimitPolicies(limits []v1alpha1.RateLimit) PolicyChain {
+	var fixedLimiters []FixedWindowRateLimiter
+	var leakyLimiters []LeakyBucketRateLimiter
+	var connLimiters []ConnectionRateLimiter
+
+	for _, limit := range limits {
+		rl, err := parseRateLimit(limit)
+		if err != nil {
+			log.Error(fmt.Sprintf("error processing rate limit rule for %s - %s", limit.Name, err.Error()))
+			continue
 		}
 
+		switch rl.(type) {
+		case FixedWindowRateLimiter:
+			fwRl := rl.(FixedWindowRateLimiter)
+			fixedLimiters = append(fixedLimiters, fwRl)
+		case LeakyBucketRateLimiter:
+			lbRl := rl.(LeakyBucketRateLimiter)
+			leakyLimiters = append(leakyLimiters, lbRl)
+		case ConnectionRateLimiter:
+			crRl := rl.(ConnectionRateLimiter)
+			connLimiters = append(connLimiters, crRl)
+		default:
+			log.Errorf("unknown rate limit type %T. ignoring", rl)
+			continue
+		}
 	}
-	return fixedLimitRules
+
+	var config PolicyChainConfiguration
+	if len(fixedLimiters) > 0 {
+		config.FixedWindowLimiters = &fixedLimiters
+	}
+	if len(leakyLimiters) > 0 {
+		config.LeakyBucketLimiters = &leakyLimiters
+	}
+	if len(connLimiters) > 0 {
+		config.ConnectionLimiters = &connLimiters
+	}
+
+	return PolicyChain{"apicast.policy.rate_limit", config}
 }
 
-func contains(serviceNames []string, svc string) bool {
-	for _, service := range serviceNames {
-		if service == svc {
-			return true
+func parseRateLimit(rl v1alpha1.RateLimit) (interface{}, error) {
+	key := LimiterKey{rl.Name, "plain", "service"}
+	if rl.Source != "" {
+		switch rl.Source {
+		//TODO - Add support for more sources here, jwt etc ..
+		case "ip":
```

Comment:
I'm not so sure. What if you want to do ip per endpoint?

Maybe this would need to be an array and we would concat those operations into one key.

Reply:
Not sure I understand what you mean here.

Reply:
What if you want to limit each ip on each endpoint independently ?

You'd have global rule, where the key is: `{{ ip }}_{{ method }}_{{ path }}`. How you'd compose that key with this?